nomad/acl.go

```diff
@@ -2,15 +2,117 @@ package nomad
 
 import (
 	"fmt"
+	"net"
 	"time"
 
 	metrics "github.com/armon/go-metrics"
 	lru "github.com/hashicorp/golang-lru"
 	"github.com/hashicorp/nomad/acl"
+	"github.com/hashicorp/nomad/helper"
 	"github.com/hashicorp/nomad/nomad/state"
 	"github.com/hashicorp/nomad/nomad/structs"
 )
 
+// Authenticate extracts an AuthenticatedIdentity from the request context or
+// provided token. The caller can extract an acl.ACL, WorkloadIdentity, or other
+// identifying token to use for authorization.
+func (s *Server) Authenticate(ctx *RPCContext, secretID string) (*structs.AuthenticatedIdentity, error) {
+	// get the user ACLToken or anonymous token
+	aclToken, err := s.ResolveSecretToken(secretID)
+	identity := &structs.AuthenticatedIdentity{ACLToken: aclToken}
+
+	switch err {
+	case nil:
+		// missing or anonymous ACL will fall-through for further annotation
+		// based on the RPC context
+		if aclToken == nil || aclToken != structs.AnonymousACLToken {
+			return identity, nil
+		}
+
+	case structs.ErrTokenExpired:
+		return nil, err
+
+	case structs.ErrTokenInvalid:
+		// if it's not a UUID it might be an identity claim
+		claims, err := s.VerifyClaim(secretID)
+		if err != nil {
+			return nil, err
+		}
+		// unlike the state queries, errors here are invalid tokens
+		identity.Claims = claims
+		return identity, nil
+
+	case structs.ErrTokenNotFound:
+		// Check if the secret ID is the leader's secret ID, in which case treat
+		// it as a management token and record the server's node ID
+		leaderAcl := s.getLeaderAcl()
+		if leaderAcl != "" && secretID == leaderAcl {
+			identity.ACLToken = structs.LeaderACLToken
+		}
+
```

Comment:
going from the comment, do we want to bail here if the secretID is _not_ the leader's secretID?

Reply:
If it's not, we need to fallthru to check if the secret ID is a node secret instead. I missed that we should be breaking out of the switch here though so that we don't bother checking the node secret if we _do_ have the leader ACL.

Reply:
The two blocks inside this case are very different, so it made for a weird helper. I've fixed that by just using an `else` block and that seems less subtle to trace.